Mathlib/Algebra/Order/SuccPred.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios, Yaël Dillies
+-/
+import Mathlib.Algebra.Group.Basic
+import Mathlib.Algebra.Order.ZeroLEOne
+import Mathlib.Data.Int.Cast.Defs
+import Mathlib.Order.SuccPred.Basic
+
+/-!
+# Interaction between successors and arithmetic
+
+We define the `SuccAddOrder` and `PredSubOrder` typeclasses, for orders satisfying `succ x = x + 1`
+and `pred x = x - 1` respectively. This allows us to transfer the API for successors and
+predecessors into these common arithmetical forms.
+
+## Todo
+
+In the future, we will make `x + 1` and `x - 1` the `simp`-normal forms for `succ x` and `pred x`
+respectively. This will require a refactor of `Ordinal` first, as the `simp`-normal form is
+currently set the other way around.
+-/
+
+/-- A typeclass for `succ x = x + 1`. -/
+class SuccAddOrder (α : Type*) [Preorder α] [Add α] [One α] extends SuccOrder α where
+  succ_eq_add_one (x : α) : succ x = x + 1
+
+/-- A typeclass for `pred x = x - 1`. -/
+class PredSubOrder (α : Type*) [Preorder α] [Sub α] [One α] extends PredOrder α where
+  pred_eq_sub_one (x : α) : pred x = x - 1
+
+variable {α : Type*} {x y : α}
+
+namespace Order
+
+section Preorder
+
+variable [Preorder α]
+
+section Add
+
+variable [Add α] [One α] [SuccAddOrder α]
+
+theorem succ_eq_add_one (x : α) : succ x = x + 1 :=
+  SuccAddOrder.succ_eq_add_one x
+
+theorem add_one_le_of_lt (h : x < y) : x + 1 ≤ y := by
+  rw [← succ_eq_add_one]
+  exact succ_le_of_lt h
+
+theorem add_one_le_iff_of_not_isMax (hx : ¬ IsMax x) : x + 1 ≤ y ↔ x < y := by
+  rw [← succ_eq_add_one, succ_le_iff_of_not_isMax hx]
+
+theorem add_one_le_iff [NoMaxOrder α] : x + 1 ≤ y ↔ x < y :=
+  add_one_le_iff_of_not_isMax (not_isMax x)
+
+@[simp]
+theorem wcovBy_add_one (x : α) : x ⩿ x + 1 := by
+  rw [← succ_eq_add_one]
+  exact wcovBy_succ x
+
+@[simp]
+theorem covBy_add_one [NoMaxOrder α] (x : α) : x ⋖ x + 1 := by
+  rw [← succ_eq_add_one]
+  exact covBy_succ x
+
+end Add
+
+section Sub
+
+variable [Sub α] [One α] [PredSubOrder α]
+
+theorem pred_eq_sub_one (x : α) : pred x = x - 1 :=
+  PredSubOrder.pred_eq_sub_one x
+
+theorem le_sub_one_of_lt (h : x < y) : x ≤ y - 1 := by
+  rw [← pred_eq_sub_one]
+  exact le_pred_of_lt h
+
+theorem le_sub_one_iff_of_not_isMin (hy : ¬ IsMin y) : x ≤ y - 1 ↔ x < y := by
+  rw [← pred_eq_sub_one, le_pred_iff_of_not_isMin hy]
+
+theorem le_sub_one_iff [NoMinOrder α] : x ≤ y - 1 ↔ x < y :=
+  le_sub_one_iff_of_not_isMin (not_isMin y)
+
+@[simp]
+theorem sub_one_wcovBy (x : α) : x - 1 ⩿ x := by
+  rw [← pred_eq_sub_one]
+  exact pred_wcovBy x
+
+@[simp]
+theorem sub_one_covBy [NoMinOrder α] (x : α) : x - 1 ⋖ x := by
+  rw [← pred_eq_sub_one]
+  exact pred_covBy x
+
+end Sub
+
+@[simp]
+theorem succ_zero [AddZeroClass α] [One α] [SuccAddOrder α] : succ (0 : α) = 1 := by
+  rw [succ_eq_add_one, zero_add]
```

Comment:
I still believe you can get rid of this lemma...

Reply:
Why do we want to get rid of it? It fits the design of "use arithmetic instead of `succ`" and it's a convenient shortcut for `succ_eq_add_one + zero_add`.

Reply:
> It fits the design of "use arithmetic instead of `succ`" and it's a convenient shortcut for `succ_eq_add_one + zero_add`.

Well clearly it doesn't since it uses `succ`! Once you de`succ`-ify it, you get `(0 : α) + 1 = 1`, which is a purely algebraic lemma

Reply:
Alright, I'll remove it.